### ISSUE TITLE
Bug: Provider context value not memoized causes ref churn in HlsVideo / DashVideo / etc., rewinding video at end of stream

### ISSUE THREAD

**Issue by akantic** [videojs/v10#1296]
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Which package(s) or projects are affected?

React (@videojs/react)

### Browser/OS/Node environment

@videojs/react@10.0.0-beta.15
@videojs/core@10.0.0-beta.15
hls.js@1.6.15
react@19, Next.js 16, Chrome 142

## Summary

`createPlayer()`'s `Provider` builds a fresh context value object on every render, and `attachMediaElement()` returns a fresh callback ref on every call. Together these cause `<HlsVideo>`'s `<video>` element ref to be detached and re-attached on every parent re-render, which calls `hls.detachMedia()` / `hls.attachMedia()` and resets the MSE source. **At end of stream this rewinds the video to `t=0` and plays it again.**

### 1. Provider context value is rebuilt on every render

[`packages/react/src/player/create-player.tsx`](https://github.com/videojs/v10/blob/main/packages/react/src/player/create-player.tsx)

```tsx
return (
  <PlayerContextProvider value={{ store, media, setMedia, container, setContainer }}>
    {children}
  </PlayerContextProvider>
);
```

Every render of `Provider` produces a new `value` object identity, so every consumer of `useContext(PlayerContext)` re-renders — including `useMediaAttach()`, `useMedia()`, `useContainer()`, `useContainerAttach()`. This was made strictly worse in [#975](https://github.com/videojs/v10/pull/975) when `setMedia` and `setContainer` were added to the context.

### 2. `attachMediaElement()` returns a new callback ref every call

[`packages/react/src/utils/attach-media-element.ts`](https://github.com/videojs/v10/blob/main/packages/react/src/utils/attach-media-element.ts)

```ts
export function attachMediaElement(media) {
  return (element) => {
    if (element) media.attach(element);
    else media.detach();
    return () => media.detach();
  };
}
```

[`HlsVideo`](https://github.com/videojs/v10/blob/main/packages/react/src/media/hls-video/index.tsx) (and `DashVideo`, `MuxVideo`, `MuxAudio`, `SimpleHlsVideo`, `NativeHlsVideo`) call this directly inside the render body:

```tsx
const composedRef = useComposedRefs(attachMediaElement(mediaApi), ref);
```

So every render produces a new callback ref. React's callback ref protocol then calls the previous ref with `null` (→ `media.detach()` → `hls.detachMedia()`) and the new ref with the element (→ `media.attach()` → `hls.attachMedia()`).

### 3. The chain at end of stream

1. Video reaches end → `onEnded` fires → app state updates → parent re-renders
2. Re-render reaches `Provider` → new context `value` object
3. All `useContext(PlayerContext)` consumers re-render → `HlsVideo` re-renders
4. `attachMediaElement(mediaApi)` returns a new ref function
5. React calls old ref with `null` → `hls.detachMedia()`
6. React calls new ref with `<video>` → `hls.attachMedia(<video>)` → fresh MSE
7. hls.js processes `BUFFER_RESET` → `startLoad(0)` → video plays from frame 0

### Stack trace

```
at hls.detachMedia
at HlsJsMediaDelegate.detach
at HlsMediaDelegate.detach
at HlsMedia.detach
at attach-media-element.js          ← the callback ref
at use-composed-refs.js
at safelyDetachRef                   ← React detaching the ref
at commitMutationEffectsOnFiber
```

Followed shortly by `safelyAttachRef → commitLayoutEffectOnFiber → hls.attachMedia → hls.startLoad(0)`.

## Proposed fix

Two complementary changes:

1. **Wrap the `Provider` context value in `useMemo`** in `create-player.tsx`, keyed on its five members. This is the architectural fix — it stops every `useContext(PlayerContext)` consumer from re-rendering on parent re-renders.

2. **Cache `attachMediaElement(media)` per media instance** via a `WeakMap` keyed on the media object. This is defense in depth — even if a media component re-renders for some other reason in the future, the callback ref identity stays stable for the lifetime of that media instance, so React doesn't churn the `<video>` ref. `WeakMap` lets the entry be GC'd when the media instance is destroyed.

Verified locally via `pnpm patch`: before, end-of-stream produced `evt:ended → RENDER → hls.detachMedia → hls.attachMedia → hls.startLoad(0)`; after, `evt:ended` produces silence and the video stays paused at the end.

## Notes

- Affects all 6 media components that use `attachMediaElement`: `HlsVideo`, `DashVideo`, `MuxVideo`, `MuxAudio`, `SimpleHlsVideo`, `NativeHlsVideo`.
- Silent during normal playback (hls.js preserves the resume position), only user-visible at end of stream.

## Reproduction

Any setup where a parent component above `<player.Provider>` re-renders during normal playback. Minimal version:

```tsx
function Parent() {
  const [done, setDone] = useState(false);
  return (
    <>
      <player.Provider>
        <player.Container>
          <HlsVideo
            src="https://test-streams.mux.dev/x36xhzz/x36xhzz.m3u8"
            onEnded={() => setDone(true)}
          />
        </player.Container>
      </player.Provider>
      {done && <button>Continue</button>}
    </>
  );
}
```

1. Play the video to the end.
2. `onEnded` fires → `setDone(true)` → Parent re-renders.
3. The video unloads and starts playing again from `t=0`.

### Current Behavior

`hls.detachMedia()` is called, then `hls.attachMedia(<video>)`, then `hls.startLoad(0)`. The video reloads and plays from the beginning.

### Expected Behavior

The video stays paused at the end of the stream. Parent re-renders should not affect playback state.

**Comment by mav2287** [comment]
Hit this bug on a React 19 / Next.js 16 project using `HlsVideo` — independently reproduced the same root cause (provider value rebuilt on every render + `attachMediaElement` returning a fresh closure on every call). Put together a fix that matches the proposed approach in the issue.

Can't open a PR directly — looks like this repo restricts PR creation to collaborators. Happy to submit a proper PR if someone can add me, or the commit is ready to cherry-pick as-is.

**Branch:** https://github.com/mav2287/v10/tree/fix/1296-provider-value-attach-ref-churn
**Commit:** [`22df048`](https://github.com/mav2287/v10/commit/22df048d961cfa1137cbc3dd30c9b47628dd7ab1)

## What the fix does

- **`packages/react/src/player/create-player.tsx`** — wraps the `Provider` context value in `useMemo` keyed on `[store, media, container]`. `setMedia` and `setContainer` are omitted because `useState` setters are identity-stable (matches the convention in `apps/sandbox/app/shell/app.tsx` where Biome's `useExhaustiveDependencies` flags them as unnecessary).

- **`packages/react/src/utils/attach-media-element.ts`** — caches the callback ref per `media` instance via a module-scoped `WeakMap<Attachable, ...>`. Repeat calls for the same media return the cached closure. `WeakMap` lets the entry be GC'd when the media instance is destroyed, so nothing leaks.

## Tests

Added `packages/react/src/utils/tests/attach-media-element.test.ts` (new file, 6 tests) and one test in `packages/react/src/player/tests/create-player.test.tsx`.

Ran the full `@videojs/react` test suite against a fresh clone:

| Step | Result |
|---|---|
| Baseline (untouched `main`) | 14 files, **200 passing** |
| With fix applied | 15 files, **207 passing** (+7 new) |
| Source reverted, new tests kept | The 2 new regression tests **fail** with `toBe` identity-comparison errors (expected — this is the "fail-before, pass-after" contract from `CONTRIBUTING.md`) |
| Source re-applied | **207/207 passing** again |
| `pnpm typecheck` | Clean |
| `pnpm lint` (on the 4 changed files) | Clean |

## Real-world verification

Also verified end-to-end against a real HLS stream served from our production CDN, in Chrome (hls.js / MSE path) and Safari (native HLS path). Both behave identically — video plays through to completion without rewinding, and parent state updates during playback no longer trigger `detachMedia` / `attachMedia` cycles.

## Diff stats

4 files changed, 176 insertions(+), 9 deletions(-) — the bulk is the new test file (87 lines). Actual production code change is 37 lines added, 5 deleted across two files.

No API changes, no type changes, no new dependencies.

**Comment by lnhrdt** [comment]
Just confirming @mav2287's patch successfully resolved the infinite re-render issue for us. In our project, we found it fixes the issue not only in HlsVideo but also in the DashVideo component.

import {HlsVideo} from '@videojs/react/media/hls-video'
import {DashVideo} from '@videojs/react/media/dash-video'

**Comment by R-Delfino95** [comment]
Hey! Thanks for reporting this. I tested it and the issue is no longer reproducible in the latest version, so I am closing it.
However, I want to include @mav2287's fix to prevent unnecessary re-renders when the Parent changes. I created [a PR](https://github.com/videojs/v10/pull/1541) for this, let's continue the discussion there. 

Thanks!
